### ISSUE TITLE
Low performance accessing to headers function using next/headers in Production

### ISSUE THREAD

**Issue by cyphersys** [vercel/next.js#58147]
### Link to the code that reproduces this issue

https://github.com/cyphersys/perf.git

### To Reproduce

1. You can reproduce this issue creating a next project using npx create-next-app@latest
2. Alternative method is cloning this Git repo https://github.com/cyphersys/perf.git
3. Run npm build && npm start
4. If you are starting the repo locally you must stress the following urls:
http://localhost:3000/no-headers
and
http://localhost:3000/with-headers

### Current vs. Expected behavior

If you stress the url http://localhost:3000/no-headers my node server instance can respond approximately 1500 requests per second just printing the minimum of html (a div).

![image](https://github.com/vercel/next.js/assets/7922658/699e892c-90dc-4470-a0ed-6bbe55e2248a)

On the other hand, if you stress the url http://localhost:3000/with-headers my node server instance can respond to 220 requests per second. With the only difference that I try to access the headers object of the request.

![image](https://github.com/vercel/next.js/assets/7922658/806b1549-e301-4b33-9d89-3459da3b0cf0)

### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
  Platform: win32
  Arch: x64
  Version: Windows Server 2022 Standard
Binaries:
  Node: 21.1.0
  npm: N/A
  Yarn: N/A
  pnpm: N/A
Relevant Packages:
  next: 14.0.1
  eslint-config-next: 14.0.1
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.2.2
Next.js Config:
  output: N/A

This tests are made on a dedicated server hosted at OVH:

Intel Xeon Silver 4214R @ 2.40 Ghz (2 CPU) x64
192 GB RAM
Windows Server 2022 Standard

Tried also same tests on a Linux Ubuntu 22.04 x64 using node 21.1.0 and same results.
```

### Which area(s) are affected? (Select all that apply)

App Router, CLI (create-next-app)

### Additional context

Current example is running on Next.js v14.0.1 using App Router.

I tried the same example on latest canary 14.0.2-canary.18 and performance is lightly worse.

**Comment by cyphersys** [comment]
Tested on latest 14.0.2, the issue persists.

**Comment by jpark011** [comment]
+1 
I am also having the same performance issue when using dynamic functions (`headers()`, `generateMetatdata()`...)

I know dynamic functions disable Next.js caching in the route, but there are times I need to render differently on the user agent (mobile vs. desktop).
How am I supposed to keep consistent performance with dynamic functions?

**Comment by devajakamerus** [comment]
Isn't that just the cost of dynamic route? The no-headers page is static so it's much faster. The rps ratio is similar with any dynamic page, no matter if it's headers, id params etc. 

Surely the time to read the header is minimal, like 0.01ms or something like that. You can crudely check it doing something like
`const ts = performance.now();
const headerList = headers();
const host = headerList.get("host");
let dt = performance.now() - ts;
console.log(dt); or print in page`

**Comment by cyphersys** [comment]
> Isn't that just the cost of dynamic route? The no-headers page is static so it's much faster. The rps ratio is similar with any dynamic page, no matter if it's headers, id params etc.
> 
> Surely the time to read the header is minimal, like 0.01ms or something like that. You can crudely check it doing something like `const ts = performance.now(); const headerList = headers(); const host = headerList.get("host"); let dt = performance.now() - ts; console.log(dt); or print in page`

I agree with you that there should be a cost for accessing a dynamic route, but it doesn't seem right to me that the cost is so high. Maybe, the high cost does not come from accessing to headers only, but from accessing any information in the request object as cookies, search params are also affected. That's why I thought the problem originated from accessing headers.

My IIS server running on the same hardware returns a dynamic page with a single LINQ operation in .NET 7.0 with a ratio of 30,000 rps. I was expecting a slightly closer ratio from Next.js for at least a static page.

Regards

**Comment by balazsorban44** [comment]
Hi, thanks for opening this issue!

Rendering static vs dynamic _does_ have a cost difference in terms of the requests that can be handled at any time.

Have a read at https://nextjs.org/docs/app/building-your-application/rendering/server-components#server-rendering-strategies and https://nextjs.org/docs/app/building-your-application/rendering/server-components#dynamic-rendering to understand what Next.js is doing in those cases.

If the response is not user but rather user-group specific (for example mobile/desktop), you can also check out [Middleware](https://nextjs.org/docs/app/building-your-application/routing/middleware) which lets you add low-overhead request-time routing to your traffic between static pages.

The team is also working on a feature (Partial Prerendering) that lets you combine dynamic and static content on the same page. It delivers the static page skeleton instantly, then streams the dynamic content in as it's being generated on the server. Learn more about it here: https://nextjs.org/learn/dashboard-app/partial-prerendering#combining-static-and-dynamic-content

As the issue is not a specific bug, I'll close it for now, but please feel free to open a new one if you have other issues or have more feedback, thanks!
